Implemented --indicator-style flag on ls.

ricardoaiglesias opened the pull request:
Implemented the following options to `ls`.
-   `--indicator-style=none` (default)
-    `-F, --indicator-style=classify`: append indicator (one of `*/=>@|`) to entries
-    `--file-type, --indicator-style=file-type:` the same, but does not append `*`
-    `-p, --indicator-style=slash`: only appends `/` to directories.

 Previously, only `classify` was implemented. 

Note, I don't have much experience with `clap`, so I'm not 100% sure I implemented the flag parsing correctly (though I heavily referenced the `sort` command's options). Additionally, I was unsure how to create fifo objects and sockets within the test harness, so those two symbols are currently untested.

sylvestre commented:
@tertsdiepraam are you confortable reviewing this? thanks

sylvestre commented on src/uu/ls/src/ls.rs:
```diff
@@ -1214,11 +1324,29 @@ fn display_file_name(
+        let char_opt = match config.indicator_style {
+            IndicatorStyle::Classify => sym,
+            IndicatorStyle::FileType => {
+                // Don't append an asterisk.
```
please rustfmt the code

sylvestre commented on tests/by-util/test_ls.rs:
```diff
@@ -753,6 +753,40 @@ fn test_ls_inode() {
+    for opt in options {
+        // Verify that classify and file-type both contain indicators for symlinks.
+        let result = scene.ucmd().arg(format!("--indicator-style={}", opt)).run();
+        println!("stdout = {:?}", result.stdout);
```
ditto

ricardoaiglesias replied:
Running ` cargo fmt -- --check | echo $?` at the top level of the project outputs 0 (after the most recent commit, which just added a space to one of the match arms in line 1331), which indicates it's been properly formatted (Running `rustc --version` outputs `rustc 1.50.0`, and `rustfmt 1.4.30-stable` outputs `rustfmt 1.4.30-stable` )

tertsdiepraam commented on src/uu/ls/src/ls.rs:
```diff
@@ -603,10 +648,49 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
+                    options::indicator_style::NONE,
+                    options::indicator_style::SLASH,
+                    options::indicator_style::FILE_TYPE,
+                    options::indicator_style::CLASSIFY,
```
It's not necessary to let have these override with the `options::indicator_style::*`, since those are not names of the `Arg`s. Same goes for the other options you added.

tertsdiepraam commented on src/uu/ls/src/ls.rs:
```diff
@@ -669,8 +753,25 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
+                    options::indicator_style::FILE_TYPE,
+                    options::indicator_style::CLASSIFY,
+                ])
+        )
```
Could you group the arguments, so they're all next to each other in the code?

tertsdiepraam replied:
And maybe specify the possible values (with `Arg::possible_values`), so we get nice clap errors if an invalid value is given.

ricardoaiglesias replied:
Sure thing! Thanks!

tertsdiepraam reviewed (changes requested):
Looks good to me! Just one thing about the overrides and grouping the arguments.

sylvestre commented:
Would it be possible to add a bit more tests?
I see that some cases aren't covered by tests. thanks

ricardoaiglesias commented:
@sylvestre Sure! I pointed out above that I wasn't sure how to create temporary FIFO or socket objects. Which structs + methods should I be using to create these temporary objects? 

I can also add test cases for the argument parsing part of the application that isn't currently tested.

sylvestre commented:
You can do pipe this way:
https://github.com/uutils/coreutils/blob/74a5ed15443cf79d7c032f502ad26f0166a10f32/tests/by-util/test_tr.rs#L7
sockets this way:
https://github.com/uutils/coreutils/blob/cb6871d2a0bc062a0cea0d2a96762303bd6486f7/tests/by-util/test_cat.rs#L152

ricardoaiglesias commented:
After some digging, it seems that the test I'm looking for (Ensuring that `--classify` results in appending a `|` character to a FIFO) needs a named pipe, not to pipe in data from another process (since piping in data doesn't create a new file). What I found is that Rust's libc has a `mkfifo` function ([mkfifo](https://docs.rs/libc/0.2.65/libc/fn.mkfifo.html)). Should I also add a `mkfifo` wrapper in the AtPath struct (which would look like this: 
````
    pub fn mkfifo(&self, fifo: &str) {
        log_info("mkfifo", self.plus_as_string(fifo));
        libc::mkfifo(CString::new(fifo), libc::S_IWUSR | libc::S_IRUSR);
    }
````
, or should I pass on this for now and simply focus on getting coverage?

sylvestre commented:
Adding fifo tests is great, thanks

sylvestre reviewed (changes requested):
waiting for a bit more tests

ricardoaiglesias commented:
Apologies. It took a bit more time because I got caught up with work, and also needed to modify util.rs to add support for FIFOs. 

Also, for future reference, should I add `Cargo.lock` to my commits? If not, I can resolve the merge conflicts by undoing my changes.

ricardoaiglesias commented:
Just noticed that some tests failed to build on Windows because I'm using `libc`... Should I mark the relevant functions (`mkfifo` and `is_fifo`) as `#[cfg(unix)]`? I'm not sure what else could cause these tests to not build properly.

sylvestre commented:
yeah, !windows is fine

sylvestre reviewed (changes requested):
Please fix the various warnings:
https://github.com/uutils/coreutils/pull/1907/files

ricardoaiglesias commented:
All the style violations seemed to stem from an extraneous newline. Fixed that, and also added an `indicator_style` test for non-unix OSes. Hopefully everything's good now!
